Meta: Add basic clang-tidy configuration and start fixing issues in AK and Kernel.

ADKaster opened the pull request:
Add a basic clang-tidy configuration that enables checks from the following categories:

- bugprone
- cert
- clang-analyzer
- concurrency
- misc
- performance
- portability
- readability

The noisiest rules that have conflicts with the project style guide or
accepted practices have been turned off.

There's absolutely more work to be done here before we could consider
setting any of these warnings as errors and enforcing them in CI, but
committing a project clang-tidy configuration should help the rules
become more visible and let other contributors take a crack at tweaking
rules and/or finding possible bugs.

Sadly the cpp-core-guidelines and modernize categories are very, very
noisy. If we want to enable rules from these categories, they would need
to be on a rule by rule basis.

-----

Begin fixing issues in AK and Kernel from running
```
./Toolchain/Local/clang/bin/clang-tidy -p ./Build/x86_64clang Kernel/<file>.cpp
```
For `<file>` as Thread.cpp and Process.cpp

After rebuilding the Clang toolchain for x86_64.

All of the fixes seem like pretty good stylistic (as in, align older files to project style) or bug-avoidance issues.

One point of consternation might by the commits about `readability-implicit-bool-conversion`. Let me know if I should suppress that rule instead of fixing it globally.

awesomekling reviewed (commented):
Typo, `readability-qualfied-auto` -> `readability-qualified-auto`

BenWiederhake commented on AK/StringImpl.h:
```diff
@@ -60,7 +60,7 @@ class StringImpl : public RefCounted<StringImpl> {
         if (length() != other.length())
             return false;
-        return !__builtin_memcmp(characters(), other.characters(), length());
+        return (0 == __builtin_memcmp(characters(), other.characters(), length()));
```
I think the outermost parens are unnecessary.

BenWiederhake commented on AK/StringView.h:
```diff
@@ -201,7 +201,7 @@ class StringView {
         if (length() != other.length())
             return false;
-        return !__builtin_memcmp(m_characters, other.m_characters, m_length);
+        return (0 == __builtin_memcmp(m_characters, other.m_characters, m_length));
```
Again outermost parens

BenWiederhake commented on AK/Time.h:
```diff
@@ -163,7 +163,7 @@ class Time {
     [[nodiscard]] timeval to_timeval() const;
 
-    [[nodiscard]] bool is_zero() const { return !m_seconds && !m_nanoseconds; }
+    [[nodiscard]] bool is_zero() const { return (m_seconds == 0) && (m_nanoseconds == 0); }
```
Yup, I agree, this is more readable :)

BenWiederhake commented on AK/Weakable.h:
```diff
@@ -66,7 +66,7 @@ class WeakLink : public RefCounted<WeakLink> {
     bool is_null() const
     {
-        return !unsafe_ptr<void>();
+        return nullptr == unsafe_ptr<void>();
```
Although I like Yoda conditions, I think it's closer to "the common project style" to write `unsafe_ptr<void>() == nullptr`.

BenWiederhake commented on AK/StdLibExtras.h:
```diff
@@ -141,20 +141,21 @@ constexpr bool is_constant_evaluated()
+__DEFINE_GENERIC_ABS(long, 0L, labs);
+__DEFINE_GENERIC_ABS(long long, 0LL, llabs);
 #ifndef KERNEL
-__DEFINE_GENERIC_ABS(float, 0.0f, fabsf);
+__DEFINE_GENERIC_ABS(float, 0.0F, fabsf);
 __DEFINE_GENERIC_ABS(double, 0.0, fabs);
-__DEFINE_GENERIC_ABS(long double, 0.0l, fabsl);
+__DEFINE_GENERIC_ABS(long double, 0.0L, fabsl);
```
These changes look like they should belong to a separate "fix literal suffixes" commit, instead of this "Avoid else after return" commit.

dascandy replied:
All of these are clang-tidy fixes of various types; most of the time you fix warnings you'll find one type that dominates most others, but splitting them up is a lot of work for the submitter.

ADKaster replied:
I spent a lot of time with the S key in git add -p 🙃. I can split this up if it makes sense though, it is kinda different from the other fix in this file

BenWiederhake commented on AK/Atomic.h:
```diff
@@ -187,6 +187,7 @@ class Atomic {
         return *ret;
     }
 
+    // NOLINTNEXTLINE(misc-unconventional-assign-operator) Returning an object of type Atomic volatile& here is a compilation error
```
Here and the other two in this file:
`volatile` marks the method as "can be called on instances of this class, as long as the instance can be converted to volatile Atomic<T>". That is always the case, so I don't understand what `volatile` is doing here. (Really, I don't understand. It might be correct, it might be incorrect.)

My point is: It doesn't refer to the return type. The comment is therefore misleading.

I agree that we want to return a `T` here, and not an `Atomic<T>`.

Suggestion: Change the comment to say something like `We want operator= to exchange the value, so returning an object of type Atomic& here does not make sense.`

ADKaster replied:
It doesn't compile with Atomic& on the first pass of the compiler. Because it's a volatile method, it would need to be Atomic volatile &. But that construct doesn't compile at the point of use (in tests). But yes, I agree that the statement of why the written behavior is correct is better than explaining a compile error 😅

BenWiederhake commented on AK/Checked.h:
```diff
@@ -138,7 +138,8 @@ class Checked {
     constexpr Checked& operator=(U value)
     {
-        return *this = Checked(value);
+        *this = Checked(value);
+        return *this;
```
Question: What's the difference?
(I'm just curious.)

dascandy replied:
No assignment in return statement.

BenWiederhake replied:
Resolved (I can't mark it myself)

dascandy commented on AK/StringImpl.h:
```diff
@@ -60,7 +60,7 @@ class StringImpl : public RefCounted<StringImpl> {
         if (length() != other.length())
             return false;
-        return !__builtin_memcmp(characters(), other.characters(), length());
+        return (0 == __builtin_memcmp(characters(), other.characters(), length()));
```
Does clang-tidy require yoda conditions? If not, could we not use them?

ADKaster replied:
Nah these were manually added without regard for consistency. I can make them consistent and non-Yoda. But for some cases if there's a huge expression we're comparing vs 0 it seems more readable to me to put the 0 up front 🤷‍♂️

dascandy commented on AK/Time.h:
```diff
@@ -50,7 +50,7 @@ inline bool is_leap_year(int year)
 inline int days_in_year(int year)
 {
-    return 365 + is_leap_year(year);
+    return 365 + static_cast<int>(is_leap_year(year));
```
Would prefer ` ? 1 : 0` instead of the int cast for readability.

BenWiederhake commented on AK/Format.h:
```diff
@@ -87,30 +87,40 @@ struct TypeErasedParameter {
+            size_value = static_cast<i64>(*static_cast<u32 const*>(value));
+            break;
+        case Type::UInt64:
+            size_value = static_cast<i64>(*static_cast<u64 const*>(value));
```
Isn't it even undefined behavior, due to signed overflow?

Bad suggestion, but still better than UB: `size_t size_value; i64 large_signed;`, all unsigned branches just assign directly to `size_value`, all signed branches do `large_signed = static_cast<>(); VERIFY(large_signed > 0); size_value = large_signed;`
After all, you already have to treat signed and unsigned quite differently.

dascandy replied:
I'm not 100% on the specifics of converting to signed with static_cast.

BenWiederhake replied:
Static_cast here is like "conversion", by rule 8, point 2: https://en.cppreference.com/w/cpp/language/static_cast

Converting to signed with overflow is … wow! Modulo a power of two, and not undefined behavior! https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_conversions

Today I learned something: This code is actually correct, and it's neither UB nor truncation!

dascandy replied:
Can we make it more obviously correct?

trflynn89 replied:
What about something like: (warning: completely untested)
```c++
    template<typename Visitor>
    constexpr auto visit(Visitor&& visitor) const
    {
        switch (type) {
        case TypeErasedParameter::Type::UInt8:
            return visitor(*static_cast<const u8*>(value));
        case TypeErasedParameter::Type::UInt16:
            return visitor(*static_cast<const u16*>(value));
        case TypeErasedParameter::Type::UInt32:
            return visitor(*static_cast<const u32*>(value));
        case TypeErasedParameter::Type::UInt64:
            return visitor(*static_cast<const u64*>(value));
        case TypeErasedParameter::Type::Int8:
            return visitor(*static_cast<const i8*>(value));
        case TypeErasedParameter::Type::Int16:
            return visitor(*static_cast<const i16*>(value));
        case TypeErasedParameter::Type::Int32:
            return visitor(*static_cast<const i32*>(value));
        case TypeErasedParameter::Type::Int64:
            return visitor(*static_cast<const i64*>(value));
        default:
            TODO();
        }
    }

    constexpr size_t to_size() const
    {
        return visit([](auto value) {
            if constexpr (IsSigned<decltype(value)>)
                VERIFY(value > 0)
            return static_cast<std::size_t>(value);
        });
    }
```

trflynn89 replied:
Though we also probably want to check that any `u64` / `i64` parameter isn't larger than `NumericLimits<size_t>::max` on 32-bit

BenWiederhake replied:
Using that Visitor pattern, would something like this work?:
```
if constexpr (sizeof<decltype(value)> < sizeof<size_t>)
    VERIFY(value < NumericLimits<size_t>::max)
```

trflynn89 replied:
Yep, though I think you want to flip the constexpr condition from `<` to `>`
(side note, I have not had enough coffee yet to be mentally parsing all these `<` and `>` :laughing: )

BenWiederhake replied:
Derp, yes, you're right :sweat_smile: 

Fixed my code suggestion to make Andrew's life an epsilon easier.

ADKaster replied:
So wait you guys are suggesting this?
```
    constexpr size_t to_size() const 
    {
        return visit([]<typename T>(T value) {
            if constexpr (sizeof(T) > sizeof(size_t))
                VERIFY(value < NumericLimits<size_t>::max());
            if constexpr (IsSigned<T>)
                VERIFY(value > 0);
            return static_cast<size_t>(value);
        });
    }
```
It seems to avoid all the warnings, but do we need both checks? :thinking:

ADKaster replied:
Aight well that's what I went with so we can discuss in the next round of review :D

trflynn89 replied:
I think both are good to have, yes. On x86, a `u64` value of `NumericLimits<size_t>::max() + 1` would become 0 when cast to `size_t`. Probably want to alert the author of that just as much as we'd want to alert them about a negative size.

dascandy commented on AK/Userspace.h:
```diff
@@ -20,7 +20,7 @@ class Userspace {
     Userspace() = default;
 
-    operator bool() const { return m_ptr; }
+    operator bool() const { return static_cast<bool>(m_ptr); }
```
Prefer `!= nullptr` as it states intent.

ADKaster replied:
That actually doesn't work here because the type of m_ptr is a FlatPtr in one #ifdef path and a pointer in the other.

ADKaster replied:
Actually it does work, as long as I move the definitions into the ifdef blocks below. rebase incoming...

dascandy commented on Kernel/FileSystem/File.h:
```diff
@@ -124,7 +124,7 @@ class File
     void evaluate_block_conditions()
     {
-        if (Processor::current_in_irq()) {
+        if (Processor::current_in_irq() != 0) {
```
What does the int mean that this function returns?

ADKaster replied:
It's used as a boolean in practice, but the m_in_irq field on Processor is kept as a FlatPtr for what I assume are padding/appointment purposes. Most of the fields are read using offsetof and the %gs segment register

BenWiederhake commented on AK/Time.h:
```diff
@@ -139,7 +139,7 @@ class Time {
     [[nodiscard]] static Time from_timespec(const struct timespec&);
     [[nodiscard]] static Time from_timeval(const struct timeval&);
-    [[nodiscard]] constexpr static Time min() { return Time(-0x8000'0000'0000'0000LL, 0); };
+    [[nodiscard]] constexpr static Time min() { return Time(-__INT64_MAX__ - 1LL, 0); };
```
Can we use `INT64_MIN` here? We define it, and [this reference](https://www.cplusplus.com/reference/cstdint/) says that it should always exist.

ADKaster replied:
No because we don't include the right header (stdint.h) as this file is used in the Kernel and userspace. AK/Types doesn't define such a max

BenWiederhake replied:
Ah, right!

ADKaster replied:
I'll add a comment explaining this.

BenWiederhake commented on AK/Time.h:
```diff
@@ -139,7 +139,7 @@ class Time {
-    [[nodiscard]] constexpr static Time min() { return Time(-0x8000'0000'0000'0000LL, 0); };
+    [[nodiscard]] constexpr static Time min() { return Time(-__INT64_MAX__ - 1LL, 0); };
     [[nodiscard]] constexpr static Time zero() { return Time(0, 0); };
     [[nodiscard]] constexpr static Time max() { return Time(0x7fff'ffff'ffff'ffffLL, 999'999'999); };
```
If you insist on using INT64_MAX/MIN, then you should use it here, too.

ADKaster replied:
good point, symmetry would be nice here.

dascandy commented on Kernel/FileSystem/OpenFileDescription.h:
```diff
@@ -40,8 +40,8 @@ class OpenFileDescription : public RefCounted<OpenFileDescription> {
     {
-        set_readable(options & O_RDONLY);
-        set_writable(options & O_WRONLY);
+        set_readable((options & O_RDONLY) != 0);
```
Prefer the `== O_RDONLY` variant, as it does not require the flag being checked to be exactly one bit (or the code reader to know / think about that).

ADKaster replied:
Sounds reasonable, I've converted the conditions I changed (at least in the kernel) to that style.

BenWiederhake commented on Kernel/Library/ThreadSafeRefCounted.h:
```diff
@@ -98,9 +100,9 @@ class RefCounted : public RefCountedBase {
             call_will_be_destroyed_if_present(static_cast<const T*>(this));
             delete static_cast<const T*>(this);
             return true;
-        } else if (new_ref_count == 1) {
-            call_one_ref_left_if_present(static_cast<const T*>(this));
         }
+        if (new_ref_count == 1)
+            call_one_ref_left_if_present(static_cast<const T*>(this));
```
Nitpick: wrong commit

dascandy replied:
? The comment is on line 38, this is in line 101. I think the comment is correct.

BenWiederhake replied:
This particular diff around the lines 101-105 belongs into the "Avoid else after return" commit, instead of the "Suppress clang-tidy warnings from the cert-* category" commit. I'm not talking about line 38.

ADKaster replied:
Whoops good catch. Forgot this was in this file when I was splitting commits

dascandy replied:
Apologies, my mind read "commit" as "comment"...

BenWiederhake commented on AK/Bitmap.h:
```diff
@@ -38,7 +38,7 @@ class Bitmap {
 
     [[nodiscard]] BitmapView view() { return { m_data, m_size }; }
-    [[nodiscard]] BitmapView const view() const { return { m_data, m_size }; }
+    [[nodiscard]] BitmapView view() const { return { m_data, m_size }; }
```
Out of scope, but this sounds like it really should be `const`. Also, why is `BitmapView::set()` marked `const`? This subverts the entire point of `const`!

dascandy replied:
The return object is copied to the client and is copyable, so making it const is not a useful thing to do. You'd need a ConstBitmapView to do what you're saying.

Anyway, this const did nothing.

BenWiederhake replied:
I agree that it did nothing, and I agree that implementing a ConstBitmapView is out of scope :)

dascandy commented on Kernel/Thread.cpp:
```diff
@@ -786,18 +789,16 @@ static DefaultSignalAction default_signal_action(u8 signal)
+    auto const& action = m_signal_action_data[signal];
     if (action.handler_or_sigaction.is_null())
         return default_signal_action(signal) == DefaultSignalAction::Ignore;
-    if ((sighandler_t)action.handler_or_sigaction.get() == SIG_IGN)
```
if (true) return true; return false... Lovely, haven't seen one of these in a while :-)

BenWiederhake commented on Kernel/Library/ThreadSafeRefPtr.h:
```diff
@@ -49,7 +49,7 @@ struct RefPtrTraits {
     ALWAYS_INLINE static bool is_null(FlatPtr bits)
     {
-        return !(bits & ~(FlatPtr)1);
+        return 0 == (bits & ~(FlatPtr)1);
```
Here and again in `lock()`: Why the Yoda condition?

ADKaster replied:
I won't reply to all these but the reason is personal style and what came naturally when fixing them 🤷‍♂️. Will make them consistent

BenWiederhake replied:
Sure, I just flagged them all in the hope of making it easier for you to find them.

Again, I actually prefer Yoda-conditions, I just think it's not the usual project style. So in the places that you disagree (like after "huge expressions" as you said), go ahead and do a Yoda comparison.

BenWiederhake commented on Kernel/TTY/TTY.h:
```diff
@@ -41,10 +41,10 @@ class TTY : public CharacterDevice {
-    bool should_flush_on_signal() const { return !(m_termios.c_lflag & NOFLSH); }
-    bool should_echo_input() const { return m_termios.c_lflag & ECHO; }
-    bool in_canonical_mode() const { return m_termios.c_lflag & ICANON; }
+    bool should_generate_signals() const { return ISIG == (m_termios.c_lflag & ISIG); }
+    bool should_flush_on_signal() const { return 0 == (m_termios.c_lflag & NOFLSH); }
+    bool should_echo_input() const { return ECHO == (m_termios.c_lflag & ECHO); }
+    bool in_canonical_mode() const { return ICANON == (m_termios.c_lflag & ICANON); }
```
More Yoda conditions (To reiterate: I like them, but so far Serenity doesn't typically use them, so let's try to avoid them to keep the code uniform and non-surprising.)

BenWiederhake commented on Kernel/Thread.cpp:
```diff
@@ -633,11 +634,11 @@ void Thread::send_signal(u8 signal, [[maybe_unused]] Process* sender)
 
     m_pending_signals |= 1 << (signal - 1);
-    m_have_any_unmasked_pending_signals.store(pending_signals_for_state() & ~m_signal_mask, AK::memory_order_release);
+    m_have_any_unmasked_pending_signals.store(0 != (pending_signals_for_state() & ~m_signal_mask), AK::memory_order_release);
```
More Yoda (I counted 6 more in this file in this commit)

BenWiederhake commented on Kernel/ThreadTracer.h:
```diff
@@ -18,7 +18,7 @@ class ThreadTracer {
 
     ProcessID tracer_pid() const { return m_tracer_pid; }
-    bool has_pending_signal(u32 signal) const { return m_pending_signals & (1 << (signal - 1)); }
+    bool has_pending_signal(u32 signal) const { return 0 != (m_pending_signals & (1 << (signal - 1))); }
```
Yoda

BenWiederhake reviewed (changes requested):
Awesome! Looks like clang-tidy has turned up several true positives. Good job :D

I'm mostly complaining about Yoda conditions. I feel like Serenity usually does conditions by checking `complicate_expression() == constant_or_literal`, so let's try to stick to that style. If you disagree, feel free to disregard those comments.

BertalanD commented on Kernel/PhysicalAddress.h:
```diff
@@ -38,8 +38,9 @@ class PhysicalAddress {
 
     [[nodiscard]] bool is_null() const { return m_address == 0; }
 
+    // NOLINTNEXTLINE(readability-make-member-function-const) Non-const PhysicalAddress shouldn't be allowed to modify the underlying memory
```
```suggestion
    // NOLINTNEXTLINE(readability-make-member-function-const) Const PhysicalAddress shouldn't be allowed to modify the underlying memory
```

ADKaster replied:
Whoops, got these backwards, thanks!

BertalanD commented on Kernel/VirtualAddress.h:
```diff
@@ -37,6 +37,7 @@ class VirtualAddress {
     bool operator==(const VirtualAddress& other) const { return m_address == other.m_address; }
     bool operator!=(const VirtualAddress& other) const { return m_address != other.m_address; }
 
+    // NOLINTNEXTLINE(readability-make-member-function-const) Non-const VirtualAddress shouldn't be allowed to modify the underlying memory
```
```suggestion
    // NOLINTNEXTLINE(readability-make-member-function-const) const VirtualAddress shouldn't be allowed to modify the underlying memory
```

ADKaster commented:
Updates since last push:

- Fix typo in commit message (dang it!)
- Removed, Yoda is.
- Reword some poorly worded comments per suggestions
- Split out mis-allocated changes into their proper commits

Still some comments left to address.

ADKaster commented:
Changes since last update:

- Rebase and fix merge conflicts
- Implement last few comments
    - ... except for the one about Process::current_in_irq(). I'm not sure why this guy has to be a FlatPtr in the Process object? Seems beyond my kernel knowledge

Should be ready for final review / marge :)
